### PR TITLE
junit 5 compatibility for MockWebServer

### PR DESCRIPTION
**Pull request by yschimke**
Junit 5 support for MockWebServer via extending a known base class.

http://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4-rulesupport

### CODE REVIEW

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java [hunk 1]
```diff
@@ -398,6 +379,14 @@ public synchronized void shutdown() throws IOException {
     }
   }
 
+  public synchronized void after() {
```
Comment (swankjesse) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java]: protected

**Comment by swankjesse** [comment]
What the heck is wrong with the JUnit API designers. Breaking APIs like this is not a thing.

**Comment by JakeWharton** [comment]
This is a breaking API changes so we can't merge until OkHttp 4.x. And even when that happens, I'd prefer separate artifacts for JUnit 4 and 5. Making MockWebServer depend directly on JUnit 4 was a huge mistake on my part.

**Comment by swankjesse** [comment]
I don’t think this is a breaking change in OkHttp. ExternalResource still implements TestRule.

**Comment by JakeWharton** [comment]
Ah, cool.

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java [hunk 2]
```diff
@@ -142,7 +140,7 @@
 
   private boolean started;
 
-  private synchronized void maybeStart() {
+  protected synchronized void before() {
```
Comment (JakeWharton) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java]: Missing `@Override`?

File: mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java (hunk 1 above)
Comment (JakeWharton) [inline mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.java]: Missing `@Override`?

**Comment by yschimke** [comment]
Yeah, it looks worse than it is.  ExternalResource has been around since 4.7 and just removes some boilerplate.

I tried upgrading my code to junit5 and this was the only thing I was hitting. This isn't urgent but seems fully backwards compatible.

**Comment by yschimke** [comment]
And no actual dependency on junit 5 anywhere.

**Comment by JakeWharton** [comment]
Yeah I was reviewing on mobile so I jumped to a few conclusions. This does
look good.

**Comment by swankjesse** [comment]
Thanks!

**Pull request merged**